# Data Preprocessing and Cleaning

**This is the initial stage where the raw data is prepared for analysis. The goal is to ensure the data is clean, consistent, and ready for analysis**

_____________________________________________________________________________________________________________________________________________

## Table of Contents

1) [Importing of Python Libraries and Loading of Dataset](#Importing-of-Python-Libraries-and-Loading-of-Dataset)
2) [Initial Read and Data Preprocessing](#Initial-Read-and-Data-Preprocessing)
    * [Column Definitions](#Column-Definitions)
    * [Data Type Conversion](#Date-Type-Conversion)
    * [Column Nulls](#Column-Nulls)
    * ['NSA' Values](#'NSA'-Values)
3) [Data Cleaning](#Data-Cleaning)
    * [Column Removal](#Column-Removal)
    * [Removal of Rows with Null Values](#Removal-of-Rows-with-Null-Values)
    * [Conversion of Data Types](#Conversion-of-Data-Types)
4) [Import, Clean and Merge Police Divisions Dataset](#Import,-Clean-and-Merge-Police-Divisions-Dataset)
    * [Nulls and Data Types](#Nulls-and-Data-Types)
    * [Column Renaming and Removal](#Column-Renaming-and-Removal)
    * [Division 54 Specific Nulls Analysis](#Division-54-Specific-Nulls-Analysis)
    * [Removal of D54 Values](#Removal-of-D54-Values)

## Importing of Python Libraries and Loading of Dataset

In [6]:
# Importing of Pandas library
import pandas as pd

# Loading of csv format major crime indicators dataset into pandas dataframe
mci_raw_df = pd.read_csv('Major_Crime_Indicators_Raw.csv')

## Initial Read and Data Preprocessing

In [8]:
# Adjusting print options to display all columns

pd.set_option('display.max_columns', None)

# Display first 10 rows of data

mci_raw_df.head(10)

OBJECTID EVENT_UNIQUE_ID          REPORT_DATE               OCC_DATE  \
0         1  GO-20141261609  1/1/2014 5:00:00 AM    1/1/2014 5:00:00 AM   
1         2  GO-20141260033  1/1/2014 5:00:00 AM  12/31/2013 5:00:00 AM   
2         3  GO-20141260127  1/1/2014 5:00:00 AM    1/1/2014 5:00:00 AM   
3         4  GO-20141260597  1/1/2014 5:00:00 AM    1/1/2014 5:00:00 AM   
4         5  GO-20141260618  1/1/2014 5:00:00 AM    1/1/2014 5:00:00 AM   
5         6  GO-20141260618  1/1/2014 5:00:00 AM    1/1/2014 5:00:00 AM   
6         7  GO-20141260056  1/1/2014 5:00:00 AM    1/1/2014 5:00:00 AM   
7         8  GO-20141263672  1/1/2014 5:00:00 AM    1/1/2014 5:00:00 AM   
8         9  GO-20141262608  1/1/2014 5:00:00 AM    1/1/2014 5:00:00 AM   
9        10  GO-20141263706  1/1/2014 5:00:00 AM    1/1/2014 5:00:00 AM   

   REPORT_YEAR REPORT_MONTH  REPORT_DAY  REPORT_DOY  REPORT_DOW  REPORT_HOUR  \
0         2014      January           1           1  Wednesday             9   
1         2014      January           1           1  Wednesday             2   
2         2014      January           1           1  Wednesday             1   
3         2014      January           1           1  Wednesday             2   
4         2014      January           1           1  Wednesday             5   
5         2014      January           1           1  Wednesday             5   
6         2014      January           1           1  Wednesday             1   
7         2014      January           1           1  Wednesday            18   
8         2014      January           1           1  Wednesday            14   
9         2014      January           1           1  Wednesday            18   

   OCC_YEAR OCC_MONTH  OCC_DAY  OCC_DOY     OCC_DOW  OCC_HOUR DIVISION  \
0    2014.0   January      1.0      1.0  Wednesday          9      D55   
1    2013.0  December     31.0    365.0  Tuesday           22      D42   
2    2014.0   January      1.0      1.0  Wednesday          1      D14   
3    2014.0   January      1.0      1.0  Wednesday          2      D14   
4    2014.0   January      1.0      1.0  Wednesday          2      D14   
5    2014.0   January      1.0      1.0  Wednesday          2      D14   
6    2014.0   January      1.0      1.0  Wednesday          1      D53   
7    2014.0   January      1.0      1.0  Wednesday         18      D54   
8    2014.0   January      1.0      1.0  Wednesday         14      D14   
9    2014.0   January      1.0      1.0  Wednesday         18      D12   

                                       LOCATION_TYPE PREMISES_TYPE  UCR_CODE  \
0                   Apartment (Rooming House, Condo)     Apartment      1430   
1  Single Home, House (Attach Garage, Cottage, Mo...         House      1430   
2                                   Bar / Restaurant    Commercial      1420   
3                   Apartment (Rooming House, Condo)     Apartment      1430   
4                                   Bar / Restaurant    Commercial      1430   
5                                   Bar / Restaurant    Commercial      1430   
6                                   Bar / Restaurant    Commercial      1430   
7                                   Bar / Restaurant    Commercial      1430   
8  Streets, Roads, Highways (Bicycle Path, Privat...       Outside      1430   
9  Single Home, House (Attach Garage, Cottage, Mo...         House      1430   

   UCR_EXT              OFFENCE MCI_CATEGORY HOOD_158  \
0      100              Assault      Assault      069   
1      100              Assault      Assault      144   
2      110  Assault Bodily Harm      Assault      084   
3      100              Assault      Assault      080   
4      100              Assault      Assault      081   
5      100              Assault      Assault      081   
6      100              Assault      Assault      095   
7      100              Assault      Assault      059   
8      100              Assault      Assault      095   
9      100              Assaul

In [9]:
# Display last 5 rows of data

mci_raw_df.tail(10)

OBJECTID EVENT_UNIQUE_ID           REPORT_DATE              OCC_DATE  \
408918    408919  GO-20242142574  9/30/2024 5:00:00 AM  9/30/2024 5:00:00 AM   
408919    408920  GO-20242143865  9/30/2024 5:00:00 AM  9/15/2024 5:00:00 AM   
408920    408921  GO-20242144497  9/30/2024 5:00:00 AM  9/30/2024 5:00:00 AM   
408921    408922  GO-20242139878  9/30/2024 5:00:00 AM  9/30/2024 5:00:00 AM   
408922    408923  GO-20242145346  9/30/2024 5:00:00 AM  9/30/2024 5:00:00 AM   
408923    408924  GO-20242140191  9/30/2024 5:00:00 AM  7/16/2024 5:00:00 AM   
408924    408925  GO-20242142542  9/30/2024 5:00:00 AM  9/29/2024 5:00:00 AM   
408925    408926  GO-20242140744  9/30/2024 5:00:00 AM  9/29/2024 5:00:00 AM   
408926    408927  GO-20242143481  9/30/2024 5:00:00 AM  9/30/2024 5:00:00 AM   
408927    408928  GO-20242144170  9/30/2024 5:00:00 AM  9/30/2024 5:00:00 AM   

        REPORT_YEAR REPORT_MONTH  REPORT_DAY  REPORT_DOY  REPORT_DOW  \
408918         2024    September          30         274  Monday       
408919         2024    September          30         274  Monday       
408920         2024    September          30         274  Monday       
408921         2024    September          30         274  Monday       
408922         2024    September          30         274  Monday       
408923         2024    September          30         274  Monday       
408924         2024    September          30         274  Monday       
408925         2024    September          30         274  Monday       
408926         2024    September          30         274  Monday       
408927         2024    September          30         274  Monday       

        REPORT_HOUR  OCC_YEAR  OCC_MONTH  OCC_DAY  OCC_DOY     OCC_DOW  \
408918           12    2024.0  September     30.0    274.0  Monday       
408919           15    2024.0  September     15.0    259.0  Sunday       
408920           16    2024.0  September     30.0    274.0  Monday       
408921            5    2024.0  September     30.0    274.0  Monday       
408922           18    2024.0  September     30.0    274.0  Monday       
408923            7    2024.0       July     16.0    198.0  Tuesday      
408924           12    2024.0  September     29.0    273.0  Sunday       
408925            8    2024.0  September     29.0    273.0  Sunday       
408926           14    2024.0  September     30.0    274.0  Monday       
408927           16    2024.0  September     30.0    274.0  Monday       

        OCC_HOUR DIVISION                                      LOCATION_TYPE  \
408918        12      D22  Streets, Roads, Highways (Bicycle Path, Privat...   
408919         2      D51  Streets, Roads, Highways (Bicycle Path, Privat...   
408920        14      D55  Streets, Roads, Highways (Bicycle Path, Privat...   
408921         5      D32                                 Ttc Subway Station   
408922        17      D52  Streets, Roads, Highways (Bicycle Path, Privat...   
408923        17      D32                   Apartment (Rooming House, Condo)   
408924        20      D13  Streets, Roads, Highways (Bicycle Path, Privat...   
408925        23      D13  Single Home, House (Attach Garage, Cottage, Mo...   
408926        14      D41  Streets, Roads, Highways (Bicycle Path, Privat...   
408927         7      D23  Parking Lots (Apt., Commercial Or Non-Commercial)   

       PREMISES_TYPE  UCR_CODE  UCR_EXT                 OFFENCE MCI_CATEGORY  \
408918       Outside      1610      220         Robbery - Other      Robbery   
408919       Outside      1430      100                 Assault      Assault   
408920       Outside      1610      200       Robbery - Mugging      Robbery   
408921       Transit      1430      100                 Assault      Assault   
408922       Outside      1610      220         Robbery - Other      Robbery   
408923     Apartment      1430      100                 Assault      Assault   
408924       Outside      2135      210  Theft Of Motor Vehicle   Auto Theft  

### Column Definitions

---------------------------------------------------------------------------------------------------------------------------------------------

Below are the definitions for columns that may not be immediately understandable:

- <b>"OCC"</b>: Represents the date the offense occurred. It's important to differentiate "OCC" dates from report dates, as these reflect the time gap between when the incident happened and when it was reported to the Toronto Police Service.

- <b>"DIVISION"</b>: Indicates the division where the crime occurred. The Toronto Police Service is divided into several divisions, each responsible for specific areas of the city. While this dataset does not specify the square kilometers each division covers, this information is available in another dataset provided by the Toronto Police Service. I will merge that dataset in the final steps to facilitate easier division analysis, as many community members, including myself, are curious about the relationship between the size of a division and the nature of crimes within its boundaries.

- <b>'PREMISES_TYPE'</b>: A subcategory of <b>'LOCATION_TYPE'</b>. Each Location Type consists of multiple Premises Types.

- '<b>OFFENSE</b>': A subcategory of <b>'MCI_CATEGORY'</b>. Each MCI Category is made up of multiple offences.

- <b>'UCR_CODE'</b>: Refers to Uniform Crime Reporting (UCR) Codes, which are standardized codes used by police services, including the Toronto Police Service, to classify and report different types of criminal offenses. For instance, a UCR Code might indicate an incident is classified as an assault, while the UCR Extension provides details about whether the assault occurred in a bar, on the street, or in a home. To simplify for community members who may not be familiar with UCR codes, I will exclude them from the dataset and refer directly to the MCI Category, Offense, Location Type, and Premises Type.  

- <b>'HOOD_158'</b> and <b>'NEIGHBOURHOOD_158'</b> Identifiers of neighborhoods using the City of Toronto's new 158 neighborhood structure.

- <b>'HOOD_140'</b> and <b>'NEIGHBOURHOOD_140'</b>: Identifiers using Toronto's old neighborhood structure. To simplify our analysis and maintain consistency, I will remove the old neighborhood identifiers in the next step, ensuring we have one source of truth for geographical data.

- <b>'LONG_WGS84'</b> and <b>'LAT_WGS84'</b>: Represent geographic coordinates in WGS84 (World Geodetic System 1984) format. Meanwhile, <b>'x'</b> and <b>'y'</b> represent geographic coordinates in the projected coordinate system. As our final Tableau dashboard will only accept WGS84 geographic coordinates, and to simplify our analysis, we will remove the projected coordinate system columns.

---------------------------------------------------------------------------------------------------------------------------------------------

In [11]:
# Summary of data

mci_raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 408928 entries, 0 to 408927
Data columns (total 31 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   OBJECTID           408928 non-null  int64  
 1   EVENT_UNIQUE_ID    408928 non-null  object 
 2   REPORT_DATE        408928 non-null  object 
 3   OCC_DATE           408928 non-null  object 
 4   REPORT_YEAR        408928 non-null  int64  
 5   REPORT_MONTH       408928 non-null  object 
 6   REPORT_DAY         408928 non-null  int64  
 7   REPORT_DOY         408928 non-null  int64  
 8   REPORT_DOW         408928 non-null  object 
 9   REPORT_HOUR        408928 non-null  int64  
 10  OCC_YEAR           408792 non-null  float64
 11  OCC_MONTH          408792 non-null  object 
 12  OCC_DAY            408792 non-null  float64
 13  OCC_DOY            408792 non-null  float64
 14  OCC_DOW            408792 non-null  object 
 15  OCC_HOUR           408928 non-null  int64  
 16  DI

## Data Type Conversion

---------------------------------------------------------------------------------------------------------------------------------------------

- <b>Convert Dates</b>: Full dates such as 'REPORT_DATE' and 'OCC_DATE' should be converted to date values for time-series analysis. Date values, as opposed to date-time values, will be used because the time portion for these columns is set to 5:00 AM for each value. The accurate hour of the incident for these columns is located in 'REPORT_HOUR' and 'OCC_HOUR', respectively.

- <b>Categorical Conversion</b>: 'REPORT_MONTH', 'REPORT_DOW', 'OCC_MONTH', and 'OCC_DOW' should be converted to categorical data types for better performance.

- <b>Integer Conversion</b>: 'OCC_YEAR', 'OCC_DAY', and 'OCC_DOY' should be converted to integer data types for improved performance.

---------------------------------------------------------------------------------------------------------------------------------------------

In [13]:
# Determine general distributon of null values

nan_indices = mci_raw_df[mci_raw_df.isnull().any(axis=1)].index
print(nan_indices)

print('\n')

# Display sample of 15 rows with null values, to understand nature of rows with null values

mci_raw_df.loc[nan_indices].head(15)

Index([    44,    379,   1067,   1085,   3335,   3380,   5042,   5155,   5156,
         5529,
       ...
       392993, 395416, 396318, 396360, 399922, 399996, 402070, 402116, 402783,
       404894],
      dtype='int64', length=136)




OBJECTID EVENT_UNIQUE_ID           REPORT_DATE              OCC_DATE  \
44          45  GO-20141262074   1/1/2014 5:00:00 AM   6/1/1998 5:00:00 AM   
379        380  GO-20141292170   1/6/2014 5:00:00 AM  1/31/1996 5:00:00 AM   
1067      1068  GO-20141396749  1/14/2014 5:00:00 AM   1/1/1980 5:00:00 AM   
1085      1086  GO-20141396749  1/14/2014 5:00:00 AM   1/1/1980 5:00:00 AM   
3335      3336  GO-20141514632  2/12/2014 5:00:00 AM  10/1/1999 5:00:00 AM   
3380      3381  GO-20141514632  2/12/2014 5:00:00 AM  10/1/1999 5:00:00 AM   
5042      5043  GO-20141635586   3/4/2014 5:00:00 AM   1/1/1989 5:00:00 AM   
5155      5156  GO-20141150249   3/6/2014 5:00:00 AM   1/1/1995 5:00:00 AM   
5156      5157  GO-20141150249   3/6/2014 5:00:00 AM   1/1/1995 5:00:00 AM   
5529      5530  GO-20141672344  3/10/2014 5:00:00 AM   1/1/1987 5:00:00 AM   
5531      5532  GO-20141672343  3/10/2014 5:00:00 AM   1/1/1986 5:00:00 AM   
5554      5555  GO-20141672349  3/10/2014 5:00:00 AM   1/1/1989 5:00:00 AM   
9276      9277  GO-20141947935  4/24/2014 5:00:00 AM  4/10/1978 5:00:00 AM   
9412      9413  GO-20141963354  4/26/2014 5:00:00 AM   1/1/1998 5:00:00 AM   
9427      9428  GO-20141963354  4/26/2014 5:00:00 AM   1/1/1998 5:00:00 AM   

      REPORT_YEAR REPORT_MONTH  REPORT_DAY  REPORT_DOY  REPORT_DOW  \
44           2014      January           1           1  Wednesday    
379          2014      January           6           6  Monday       
1067         2014      January          14          14  Tuesday      
1085         2014      January          14          14  Tuesday      
3335         2014     February          12          43  Wednesday    
3380         2014     February          12          43  Wednesday    
5042         2014        March           4          63  Tuesday      
5155         2014        March           6          65  Thursday     
5156         2014        March           6          65  Thursday     
5529         2014        March          10          69  Monday       
5531         2014        March          10          69  Monday       
5554         2014        March          10          69  Monday       
9276         2014        April          24         114  Thursday     
9412         2014        April          26         116  Saturday     
9427         2014        April          26         116  Saturday     

      REPORT_HOUR  OCC_YEAR OCC_MONTH  OCC_DAY  OCC_DOY OCC_DOW  OCC_HOUR  \
44             12       NaN       NaN      NaN      NaN     NaN        12   
379            13       NaN       NaN      NaN      NaN     NaN         0   
1067           14       NaN       NaN      NaN      NaN     NaN        12   
1085           14       NaN       NaN      NaN      NaN     NaN        12   
3335            8       NaN       NaN      NaN      NaN     NaN        12   
3380            8       NaN       NaN      NaN      NaN     NaN        12   
5042            1       NaN       NaN      NaN      NaN     NaN         1   
5155           16       NaN       NaN      NaN      NaN     NaN        12   
5156           16       NaN       NaN      NaN      NaN     NaN        12   
5529            0       NaN       NaN      NaN      NaN     NaN         0   
5531            0       NaN       NaN      NaN      NaN     NaN         0   
5554            0       NaN       NaN      NaN      NaN     NaN         0   
9276            8       NaN       NaN      NaN      NaN     NaN         0   
9412           15       NaN       NaN      NaN      NaN     NaN         0   
9427           15       NaN       NaN      NaN      NaN     NaN         0   

     DIVISION                                      LOCATION_TYPE  \
44        D32                   Apartment (Rooming House, Condo)   
379       D33  Single Home, House (Attach Garage, Cottage, Mo...   
1067      D33  Single Home, House (Attach Garage, Cottage, Mo...   
1085      D33  Single Home, House (Attach Garage, Cottage, Mo...   
3335      D43  Single Home, House (Attach Garage, Cottage, Mo..

In [14]:
# Display sorted sample of 15 first rows without null values. 
# The reasoning for the comparison here, as will be explained in the text box below, is that all rows with null values in 'OCC_DATE' are peculiarly dated before the year 2000.

mci_raw_df_date_converted = mci_raw_df

mci_raw_df_date_converted['OCC_DATE'] = pd.to_datetime(mci_raw_df_date_converted['OCC_DATE'])
mci_raw_df_date_converted['REPORT_DATE'] = pd.to_datetime(mci_raw_df_date_converted['REPORT_DATE'])

mci_raw_df_date_converted = mci_raw_df_date_converted.dropna().sort_values(by='OCC_DATE')
mci_raw_df_date_converted.head(15)

OBJECTID EVENT_UNIQUE_ID         REPORT_DATE            OCC_DATE  \
231185    231186  GO-20201358247 2020-07-23 05:00:00 2000-01-01 05:00:00   
300603    300604  GO-20221193722 2022-06-23 05:00:00 2000-01-01 05:00:00   
101348    101349   GO-2017183174 2017-01-29 05:00:00 2000-01-01 05:00:00   
301360    301361  GO-20221242031 2022-06-30 05:00:00 2000-01-01 05:00:00   
301424    301425  GO-20221242031 2022-06-30 05:00:00 2000-01-01 05:00:00   
110549    110550   GO-2017823263 2017-05-10 05:00:00 2000-01-01 05:00:00   
109907    109908   GO-2017773081 2017-05-02 05:00:00 2000-01-01 05:00:00   
10685      10686  GO-20142047470 2014-05-09 05:00:00 2000-01-01 05:00:00   
73022      73023   GO-2016534326 2016-03-29 05:00:00 2000-01-01 05:00:00   
101834    101835   GO-2017212274 2017-02-03 05:00:00 2000-01-01 05:00:00   
73002      73003   GO-2016534326 2016-03-29 05:00:00 2000-01-01 05:00:00   
407241    407242  GO-20242052175 2024-09-18 05:00:00 2000-01-01 05:00:00   
399476    399477  GO-20241601403 2024-07-22 05:00:00 2000-01-01 05:00:00   
399475    399476  GO-20241601403 2024-07-22 05:00:00 2000-01-01 05:00:00   
396437    396438  GO-20241405387 2024-06-28 05:00:00 2000-01-01 05:00:00   

        REPORT_YEAR REPORT_MONTH  REPORT_DAY  REPORT_DOY  REPORT_DOW  \
231185         2020         July          23         205  Thursday     
300603         2022         June          23         174  Thursday     
101348         2017      January          29          29  Sunday       
301360         2022         June          30         181  Thursday     
301424         2022         June          30         181  Thursday     
110549         2017          May          10         130  Wednesday    
109907         2017          May           2         122  Tuesday      
10685          2014          May           9         129  Friday       
73022          2016        March          29          89  Tuesday      
101834         2017     February           3          34  Friday       
73002          2016        March          29          89  Tuesday      
407241         2024    September          18         262  Wednesday    
399476         2024         July          22         204  Monday       
399475         2024         July          22         204  Monday       
396437         2024         June          28         180  Friday       

        REPORT_HOUR  OCC_YEAR OCC_MONTH  OCC_DAY  OCC_DOY     OCC_DOW  \
231185           19    2000.0   January      1.0      1.0  Saturday     
300603           19    2000.0   January      1.0      1.0  Saturday     
101348           23    2000.0   January      1.0      1.0  Saturday     
301360           12    2000.0   January      1.0      1.0  Saturday     
301424           12    2000.0   January      1.0      1.0  Saturday     
110549           15    2000.0   January      1.0      1.0  Saturday     
109907           16    2000.0   January      1.0      1.0  Saturday     
10685            20    2000.0   January      1.0      1.0  Saturday     
73022            18    2000.0   January      1.0      1.0  Saturday     
101834           11    2000.0   January      1.0      1.0  Saturday     
73002            18    2000.0   January      1.0      1.0  Saturday     
407241           22    2000.0   January      1.0      1.0  Saturday     
399476           21    2000.0   January      1.0      1.0  Saturday     
399475           21    2000.0   January      1.0      1.0  Saturday     
396437            2    2000.0   January      1.0      1.0  Saturday     

        OCC_HOUR DIVISION                                      LOCATION_TYPE  \
231185         0      D31  Single Home, House (Attach Garage, Cottage, Mo...   
300603        16      D51                   Apartment (Rooming House, Condo)   
101348        23      D42  Single Home, House (Attach Garage, Cottage, Mo...   
301360        12      D41  Single Home, House (Attach Garage, Cottage, Mo...   
301424        12      D41  Single Home, House (Attach Garage, Cottage

### Column Nulls

---------------------------------------------------------------------------------------------------------------------------------------------

The columns 'OCC_YEAR', 'OCC_MONTH', 'OCC_DAY', 'OCC_DOY', and 'OCC_DOW' contain null values. This is evident from the difference in the total entry count, which is 408928, compared to the entry count for these columns, which is 408792.

Upon comparing the dataset, sorted in ascending order by 'OCC_DATE', with and without null values, it becomes clear that rows with null values reference incidents that occurred before the year 2000. For our analysis, we will remove these rows. The rationale is that analyzing incidents from 2000 to 2024, which spans 24 years, is more than sufficient to educate everyday Torontonians about the nature of various crimes occurring in their neighborhoods in a simple and accessible manner.

---------------------------------------------------------------------------------------------------------------------------------------------


In [16]:
# Create dictionary of unique values

unique_values_dict = mci_raw_df.apply(lambda col:col.unique().tolist()).to_dict()

# Display dictionary of unique values, for columns that are likely to contain records that are not necessarily null, but may still be incorrect

keys_to_print = ['REPORT_YEAR','REPORT_MONTH','REPORT_DAY','REPORT_HOUR','OCC_YEAR','OCC_MONTH','OCC_DAY','OCC_DOW','OCC_HOUR','DIVISION','LOCATION_TYPE','PREMISES_TYPE','OFFENCE','MCI_CATEGORY','HOOD_158']

for key in keys_to_print:
    print(f'{key}:')
    for item in unique_values_dict[key]:
        print(item)
    print('\n')

REPORT_YEAR:
2014
2015
2016
2017
2018
2019
2020
2021
2022
2023
2024


REPORT_MONTH:
January
February
March
April
May
June
July
August
September
October
November
December


REPORT_DAY:
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31


REPORT_HOUR:
9
2
1
5
18
14
4
12
8
19
3
6
7
16
15
17
21
10
23
11
20
0
22
13


OCC_YEAR:
2014.0
2013.0
2012.0
nan
2003.0
2011.0
2004.0
2010.0
2009.0
2008.0
2006.0
2000.0
2005.0
2002.0
2001.0
2015.0
2007.0
2016.0
2017.0
2018.0
2019.0
2020.0
2021.0
2022.0
2023.0
2024.0


OCC_MONTH:
January
December
March
nan
June
November
May
September
July
February
October
August
April


OCC_DAY:
1.0
31.0
30.0
nan
29.0
26.0
25.0
2.0
27.0
21.0
24.0
23.0
3.0
19.0
18.0
20.0
4.0
15.0
5.0
11.0
6.0
12.0
28.0
17.0
7.0
13.0
22.0
8.0
14.0
9.0
10.0
16.0


OCC_DOW:
Wednesday 
Tuesday   
Thursday  
Monday    
nan
Sunday    
Friday    
Saturday  


OCC_HOUR:
9
22
1
2
18
14
12
8
4
3
0
7
16
17
15
23
11
19
20
21
10
6
13
5


DIVISION:
D55
D42
D14
D53
D54
D1

In [17]:
# Determine distribtion of 'NSA' Values

nsa_indices = mci_raw_df.apply(lambda row: 'NSA' in row.values, axis = 1)
print(nsa_indices[nsa_indices].index)

# Display sample of rows with 'NSA' Values

mci_raw_df.loc[nsa_indices].head(10)

Index([    10,     51,     57,     62,    132,    136,    137,    212,    228,
          240,
       ...
       408255, 408281, 408282, 408434, 408557, 408582, 408648, 408671, 408771,
       408815],
      dtype='int64', length=6480)


OBJECTID EVENT_UNIQUE_ID         REPORT_DATE            OCC_DATE  \
10         11  GO-20141260577 2014-01-01 05:00:00 2014-01-01 05:00:00   
51         52  GO-20141260973 2014-01-01 05:00:00 2014-01-01 05:00:00   
57         58  GO-20141260577 2014-01-01 05:00:00 2014-01-01 05:00:00   
62         63  GO-20141260537 2014-01-01 05:00:00 2014-01-01 05:00:00   
132       133  GO-20142001841 2014-01-02 05:00:00 2014-01-02 05:00:00   
136       137  GO-20141268287 2014-01-02 05:00:00 2013-12-30 05:00:00   
137       138  GO-20141270511 2014-01-02 05:00:00 2014-01-02 05:00:00   
212       213  GO-20141602781 2014-01-03 05:00:00 2014-01-03 05:00:00   
228       229  GO-20142552857 2014-01-03 05:00:00 2013-01-03 05:00:00   
240       241  GO-20141283439 2014-01-04 05:00:00 2014-01-04 05:00:00   

     REPORT_YEAR REPORT_MONTH  REPORT_DAY  REPORT_DOY  REPORT_DOW  \
10          2014      January           1           1  Wednesday    
51          2014      January           1           1  Wednesday    
57          2014      January           1           1  Wednesday    
62          2014      January           1           1  Wednesday    
132         2014      January           2           2  Thursday     
136         2014      January           2           2  Thursday     
137         2014      January           2           2  Thursday     
212         2014      January           3           3  Friday       
228         2014      January           3           3  Friday       
240         2014      January           4           4  Saturday     

     REPORT_HOUR  OCC_YEAR OCC_MONTH  OCC_DAY  OCC_DOY     OCC_DOW  OCC_HOUR  \
10             4    2014.0   January      1.0      1.0  Wednesday          2   
51             4    2014.0   January      1.0      1.0  Wednesday          4   
57             4    2014.0   January      1.0      1.0  Wednesday          2   
62             4    2014.0   January      1.0      1.0  Wednesday          4   
132           15    2014.0   January      2.0      2.0  Thursday          14   
136           15    2013.0  December     30.0    364.0  Monday             6   
137           23    2014.0   January      2.0      2.0  Thursday          21   
212           11    2014.0   January      3.0      3.0  Friday             9   
228           11    2013.0   January      3.0      3.0  Thursday          15   
240           23    2014.0   January      4.0      4.0  Saturday          23   

    DIVISION                                      LOCATION_TYPE PREMISES_TYPE  \
10       NSA  Streets, Roads, Highways (Bicycle Path, Privat...       Outside   
51       NSA  Streets, Roads, Highways (Bicycle Path, Privat...       Outside   
57       NSA  Streets, Roads, Highways (Bicycle Path, Privat...       Outside   
62       NSA  Streets, Roads, Highways (Bicycle Path, Privat...       Outside   
132      D42  Single Home, House (Attach Garage, Cottage, Mo...         House   
136      NSA  Other Commercial / Corporate Places (For Profi...    Commercial   
137      NSA  Streets, Roads, Highways (Bicycle Path, Privat...       Outside   
212      NSA  Bank And Other Financial Institutions (Money M...    Commercial   
228      D23  Other Non Commercial / Corporate Places (Non-P...         Other   
240      NSA                                   Bar / Restaurant    Commercial   

     UCR_CODE  UCR_EXT                        OFFENCE     MCI_CATEGORY  \
10       1610      200              Robbery - Mugging          Robbery   
51       2130      210                     Theft Over       Theft Over   
57       1610      180             Robbery - Swarming          Robbery   
62       1430      100                        Assault          Assault   
132      2120      200                            B&E  Break and Enter   
136      2120      200                            B&E  Break and Enter   
137      1610      220                Robbery - Other          Robbery   
212      1610      110  Robbery - Financial Institute          Robbery   
22

### 'NSA' Values

---------------------------------------------------------------------------------------------------------------------------------------------

An examination of the unique values in the dataset shows that 6480 rows use the term 'NSA' as a placeholder for geographic information, specifically assigning a value of zero to longitude and latitude coordinates. According to the Toronto Police Service's documentation, 'NSA' indicates that an event occurred within 5,000 meters outside the City of Toronto. Since incidents occurring near the borders of Toronto are of interest to concerned community members, we will not remove these records. Instead, they will be filtered out only when geographical information is required for analysis.

---------------------------------------------------------------------------------------------------------------------------------------------


## Data Cleaning

### Column Removal

In [21]:
# Setting of 'EVENT_UNIQUE_ID' as index
mci_complete_df = mci_raw_df.set_index('EVENT_UNIQUE_ID')

# Drop Unnecessary Columns
mci_complete_df = mci_complete_df.drop(['OBJECTID','UCR_CODE','UCR_EXT','HOOD_140','NEIGHBOURHOOD_140','x','y'],axis = 1)


### Removal of Rows with Null Values 

In [23]:
# Remove any row with null values
# We know from our Initial Read and Data Preprocessing stage, that only the columns 'OCC_YEAR','OCC_MONTH','OCC_DAY','OCC_DOY' and 'OCC_DOW' have null values

mci_complete_df = mci_complete_df.dropna()

# Confirm sum of null values is zero

print(mci_complete_df.isnull().sum().sum())

0


### Conversion of Data Types

In [25]:
# Convert 'REPORT_DATE' and 'OCC_DATE' to date time

mci_complete_df['REPORT_DATE'] = pd.to_datetime(mci_complete_df['REPORT_DATE']).dt.date
mci_complete_df['REPORT_DATE'] = pd.to_datetime(mci_complete_df['REPORT_DATE'])

mci_complete_df['OCC_DATE'] = pd.to_datetime(mci_complete_df['OCC_DATE']).dt.date
mci_complete_df['OCC_DATE'] = pd.to_datetime(mci_complete_df['OCC_DATE'])

# Convert 'REPORT_MONTH','REPORT_DOW','OCC_MONTH' and'OCC_DOW' to categorical

mci_complete_df[['REPORT_MONTH','REPORT_DOW','OCC_MONTH','OCC_DOW']] = mci_complete_df[['REPORT_MONTH','REPORT_DOW','OCC_MONTH','OCC_DOW']].apply(lambda col: col.astype('category'))

# Convert 'OCC_YEAR', 'OCC_DAY' and 'OCC_DOY' to integer

mci_complete_df[['OCC_YEAR','OCC_DAY','OCC_DOY']] = mci_complete_df[['OCC_YEAR','OCC_DAY','OCC_DOY']].apply(lambda col: col.astype(int))

In [26]:
# Confirm unnecessary columns have been removed and data types have been converted

print(mci_complete_df.dtypes)

REPORT_DATE          datetime64[ns]
OCC_DATE             datetime64[ns]
REPORT_YEAR                   int64
REPORT_MONTH               category
REPORT_DAY                    int64
REPORT_DOY                    int64
REPORT_DOW                 category
REPORT_HOUR                   int64
OCC_YEAR                      int64
OCC_MONTH                  category
OCC_DAY                       int64
OCC_DOY                       int64
OCC_DOW                    category
OCC_HOUR                      int64
DIVISION                     object
LOCATION_TYPE                object
PREMISES_TYPE                object
OFFENCE                      object
MCI_CATEGORY                 object
HOOD_158                     object
NEIGHBOURHOOD_158            object
LONG_WGS84                  float64
LAT_WGS84                   float64
dtype: object


## Import, Clean and Merge Police Divisions Dataset

In [28]:
# Import dataset

tps_divisions_df = pd.read_csv('tps_police_divisions.csv')

In [29]:
# Show entire dataset, as this is a relatively small table

print(tps_divisions_df)

    OBJECTID                  AGENCY  DIV    UNIT_NAME             ADDRESS  \
0          1  Toronto Police Service  D11  11 Division   2054 Davenport Rd   
1          2  Toronto Police Service  D12  12 Division    200 Trethewey Dr   
2          3  Toronto Police Service  D13  13 Division  1435 Eglinton Av W   
3          4  Toronto Police Service  D14  14 Division   350 Dovercourt Rd   
4          5  Toronto Police Service  D22  22 Division     3699 Bloor St W   
5          6  Toronto Police Service  D23  23 Division     5230 Finch Av W   
6          7  Toronto Police Service  D31  31 Division      40 Norfinch Dr   
7          8  Toronto Police Service  D32  32 Division     30 Ellerslie Av   
8          9  Toronto Police Service  D33  33 Division        50 Upjohn Rd   
9         10  Toronto Police Service  D41  41 Division  2222 Eglinton Av E   
10        11  Toronto Police Service  D42  42 Division     242 Milner Av E   
11        12  Toronto Police Service  D43  43 Division  4331 Law

In [30]:
# Display data types of columns

tps_divisions_df.dtypes

OBJECTID           int64
AGENCY            object
DIV               object
UNIT_NAME         object
ADDRESS           object
CITY              object
AREA_SQKM        float64
Shape__Area      float64
Shape__Length    float64
dtype: object

### Nulls and Data Types

---------------------------------------------------------------------------------------------------------------------------------------------

There are no null values in the dataset, and all columns are of the correct data type.

Since we are primarily using this dataset for its square kilometer column, our next step will be to remove all columns except for 'DIV' (division) and 'AREA_SQKM'. Then, we will rename the 'DIV' column to 'DIVISION' to match the MCI dataset. Finally, we will merge our original MCI dataset with this one, using the newly renamed 'DIVISION' column as the merge key.

---------------------------------------------------------------------------------------------------------------------------------------------

In [32]:
# Drop unnecessary columns

tps_divisions_df = tps_divisions_df[['DIV', 'AREA_SQKM']]

# Rename 'DIV' column to 'DIVISION'

tps_divisions_df.rename(columns = {'DIV':'DIVISION'}, inplace = True)

# Confirm columns have been dropped by sampling first 5 columns

tps_divisions_df.head()

DIVISION  AREA_SQKM
0      D11  18.678458
1      D12  24.780343
2      D13  18.601911
3      D14  14.175254
4      D22  68.121884

### Column Renaming and Removal

---------------------------------------------------------------------------------------------------------------------------------------------

Now, that our merge column has been appropriately renamed and unnecessary columns have been removed, we can merge our two datasets.

---------------------------------------------------------------------------------------------------------------------------------------------

In [34]:
# Reset index

mci_complete_df = mci_complete_df.reset_index(drop = False)

# Merge dataframes

mci_complete_df = pd.merge(mci_complete_df, tps_divisions_df, on = 'DIVISION', how = 'left')

# Re-establish 'EVENT_UNIQUE_ID' as index

mci_complete_df = mci_complete_df.set_index('EVENT_UNIQUE_ID')

# Check if merge worked by looking at first row for each unique 'DIVISION' value

group_by_division = mci_complete_df.groupby('DIVISION').first().reset_index()

group_by_division[['DIVISION','AREA_SQKM']].head(18)

DIVISION  AREA_SQKM
0       D11  18.678458
1       D12  24.780343
2       D13  18.601911
3       D14  14.175254
4       D22  68.121884
5       D23  57.642280
6       D31  42.667494
7       D32  61.060480
8       D33  52.788882
9       D41  44.575112
10      D42  84.611749
11      D43  59.370005
12      D51   8.601683
13      D52   9.228056
14      D53  31.294853
15      D54        NaN
16      D55  41.033986
17      NSA        NaN

In [35]:
# Determine if 'AREA_SQKM' null values are occuring for every instance of Division 54 (D54):

mci_complete_df[mci_complete_df['DIVISION']=='D54'][['DIVISION','AREA_SQKM']].head(10)

DIVISION  AREA_SQKM
EVENT_UNIQUE_ID                    
GO-20141263672       D54        NaN
GO-20141263744       D54        NaN
GO-20141261050       D54        NaN
GO-20141261949       D54        NaN
GO-20141262002       D54        NaN
GO-20141263536       D54        NaN
GO-20141261967       D54        NaN
GO-20141268761       D54        NaN
GO-20141268612       D54        NaN
GO-20141268762       D54        NaN

In [36]:
# Determine last occurance of Division 54 (D54)

division_54_df = mci_complete_df[mci_complete_df['DIVISION'] == 'D54']

division_54_df.groupby('DIVISION')['OCC_DATE'].max()


DIVISION
D54   2018-06-14
Name: OCC_DATE, dtype: datetime64[ns]

### Division 54 Specific Nulls Analysis

---------------------------------------------------------------------------------------------------------------------------------------------

We are encountering null values in the 'AREA_SQKM' column for Division 54 (D54). Interestingly, while the MCI dataset includes occurrences of Division 54, the Police Divisions dataset does not.

According to [this announcement](https://www.tps.ca/services/resource-centre/new-police-station-for-54-division-and-55-division-consolidated-division/) by the Toronto Police Service, Division 54 and Division 55 were consolidated. This is also evident in the dataset, as the last occurrence date of Division 54 is 2018-06-14.

To standardize the dataset, I will update all instances of D54 to D55 and adjust the null 'AREA_SQKM' values for D54 to match the 'AREA_SQKM' value for D55, which is 41.033986. This will ensure consistency and make it easier to analyze and compare data post-consolidation.

---------------------------------------------------------------------------------------------------------------------------------------------

In [38]:
# Update 'AREA_SQKM' column to change values associated with 'D54' to the area square km for 'D55'

mci_complete_df.loc[mci_complete_df['DIVISION'] == 'D54', 'AREA_SQKM'] = 41.033986

# Update 'DIVISION' column to replace 'D54' with 'D55'

mci_complete_df['DIVISION'] = mci_complete_df['DIVISION'].replace('D54','D55')

# Confirm that 'D54' is no longer in 'DIVISION' column

confirm_D54 = mci_complete_df['DIVISION'].isin(['D54']).any()
print(confirm_D54)

False


### Removal of D54 Values

---------------------------------------------------------------------------------------------------------------------------------------------

We have confirmed that the 'D54' values in the 'DIVISION' column have been removed, and the 'AREA_SQKM' values associated with 'D54' have been replaced with the 'AREA_SQKM' values associated with 'D55'.

---------------------------------------------------------------------------------------------------------------------------------------------

In [40]:
## Final confirmation of completed dataset

mci_complete_df.info()

mci_complete_df.head(10)

<class 'pandas.core.frame.DataFrame'>
Index: 408792 entries, GO-20141261609 to GO-20242144170
Data columns (total 24 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   REPORT_DATE        408792 non-null  datetime64[ns]
 1   OCC_DATE           408792 non-null  datetime64[ns]
 2   REPORT_YEAR        408792 non-null  int64         
 3   REPORT_MONTH       408792 non-null  category      
 4   REPORT_DAY         408792 non-null  int64         
 5   REPORT_DOY         408792 non-null  int64         
 6   REPORT_DOW         408792 non-null  category      
 7   REPORT_HOUR        408792 non-null  int64         
 8   OCC_YEAR           408792 non-null  int64         
 9   OCC_MONTH          408792 non-null  category      
 10  OCC_DAY            408792 non-null  int64         
 11  OCC_DOY            408792 non-null  int64         
 12  OCC_DOW            408792 non-null  category      
 13  OCC_HOUR           408792 no

REPORT_DATE   OCC_DATE  REPORT_YEAR REPORT_MONTH  REPORT_DAY  \
EVENT_UNIQUE_ID                                                                
GO-20141261609   2014-01-01 2014-01-01         2014      January           1   
GO-20141260033   2014-01-01 2013-12-31         2014      January           1   
GO-20141260127   2014-01-01 2014-01-01         2014      January           1   
GO-20141260597   2014-01-01 2014-01-01         2014      January           1   
GO-20141260618   2014-01-01 2014-01-01         2014      January           1   
GO-20141260618   2014-01-01 2014-01-01         2014      January           1   
GO-20141260056   2014-01-01 2014-01-01         2014      January           1   
GO-20141263672   2014-01-01 2014-01-01         2014      January           1   
GO-20141262608   2014-01-01 2014-01-01         2014      January           1   
GO-20141263706   2014-01-01 2014-01-01         2014      January           1   

                 REPORT_DOY  REPORT_DOW  REPORT_HOUR  OCC_YEAR OCC_MONTH  \
EVENT_UNIQUE_ID                                                            
GO-20141261609            1  Wednesday             9      2014   January   
GO-20141260033            1  Wednesday             2      2013  December   
GO-20141260127            1  Wednesday             1      2014   January   
GO-20141260597            1  Wednesday             2      2014   January   
GO-20141260618            1  Wednesday             5      2014   January   
GO-20141260618            1  Wednesday             5      2014   January   
GO-20141260056            1  Wednesday             1      2014   January   
GO-20141263672            1  Wednesday            18      2014   January   
GO-20141262608            1  Wednesday            14      2014   January   
GO-20141263706            1  Wednesday            18      2014   January   

                 OCC_DAY  OCC_DOY     OCC_DOW  OCC_HOUR DIVISION  \
EVENT_UNIQUE_ID                                                    
GO-20141261609         1        1  Wednesday          9      D55   
GO-20141260033        31      365  Tuesday           22      D42   
GO-20141260127         1        1  Wednesday          1      D14   
GO-20141260597         1        1  Wednesday          2      D14   
GO-20141260618         1        1  Wednesday          2      D14   
GO-20141260618         1        1  Wednesday          2      D14   
GO-20141260056         1        1  Wednesday          1      D53   
GO-20141263672         1        1  Wednesday         18      D55   
GO-20141262608         1        1  Wednesday         14      D14   
GO-20141263706         1        1  Wednesday         18      D12   

                                                     LOCATION_TYPE  \
EVENT_UNIQUE_ID                                                      
GO-20141261609                    Apartment (Rooming House, Condo)   
GO-20141260033   Single Home, House (Attach Garage, Cottage, Mo...   
GO-20141260127                                    Bar / Restaurant   
GO-20141260597                    Apartment (Rooming House, Condo)   
GO-20141260618                                    Bar / Restaurant   
GO-20141260618                                    Bar / Restaurant   
GO-20141260056                                    Bar / Restaurant   
GO-20141263672                                    Bar / Restaurant   
GO-20141262608   Streets, Roads, Highways (Bicycle Path, Privat...   
GO-20141263706   Single Home, House (Attach Garage, Cottage, Mo...   

                PREMISES_TYPE              OFFENCE MCI_CATEGORY HOOD_158  \
EVENT_UNIQUE_ID                                                            
GO-20141261609      Apartment              Assault      Assault      069   
GO-20141260033          House              Assault      Assault      144   
GO-20141260127     Commercial  Assault Bodily Harm      Assault      084   
GO-20141260597      Apartment              Assault      Assault      080   
GO-20141260618     Commercia

In [41]:
## Export of dataset for Exploratory Data Analysis stage

mci_complete_df.to_csv('Major_Crime_Indicators_Completed.csv')